### PR TITLE
show emojis on map

### LINKED ISSUE #1783: Emoji markers are not displayed on the map
One-letter markers are displayed directly (no need to click on the marker), but emojis don't.

Probably something is wrong with this length check and emojis length is not "1"? I tried replacing it with `neq` and `0`, but then it stopped displaying any markers at all, even 1-letter english markers:
https://github.com/deltachat/deltachat-android/blob/370335bf6e4f28a79a3f0a654035f4dafecb85ad/src/org/thoughtcrime/securesms/map/MapDataManager.java#L459

It's fixed in the core (https://github.com/deltachat/deltachat-android/blob/370335bf6e4f28a79a3f0a654035f4dafecb85ad/src/org/thoughtcrime/securesms/map/MapDataManager.java#L459), and I have tested with python that markers are returned by the FFI just as well as letters.

### PR DESCRIPTION
**Pull request by cyBerta**
closes #1783

### CODE REVIEW

**Comment by r10s** [comment]
is this still WIP?

**Comment by cyBerta** [comment]
I would like to double check that next monday

**Comment by r10s** [comment]
> I would like to double check that next monday

great!

**Comment by cyBerta** [comment]
@link2xt maybe you can have a look at this PR and try it out?

File: src/org/thoughtcrime/securesms/components/emoji/EmojiProvider.java [hunk 1]
```diff
@@ -158,6 +172,26 @@ public int getIntrinsicHeight() {
...
+      final int rowIndex = info.getIndex() % EMOJI_PER_ROW;
+
+      Rect desRect = new Rect(0, 0, (int) intrinsicWidth, (int) intrinsicWidth);
+      Rect srcRect = new Rect((int)(rowIndex * intrinsicWidth),
+        (int)(row * intrinsicHeight + row * verticalPad)+1,
+        (int)(((rowIndex + 1) * intrinsicWidth)-1),
+        (int)((row + 1) * intrinsicHeight + row * verticalPad)-1);
+
+      Canvas canvas = new Canvas(singleEmoji);
+      canvas.scale(0.5f, 0.5f);
```
Comment (link2xt) [inline src/org/thoughtcrime/securesms/components/emoji/EmojiProvider.java:188]: Where does the 0.5 scale comes from? Why the emoji returned is half the size of intrinsicWidth and intrinsicHeight?

File: src/org/thoughtcrime/securesms/map/DataCollectionTask.java [hunk 2]
```diff
@@ -64,29 +70,32 @@ public static void cancelRunningTasks() {
...
+          emojiCodePoints,
+          emojiProvider,
+          boundingBuilder);
         for (int contactId : contactIds) {
            dataCollector.updateSource(chatId,
-                    contactId,
-                    System.currentTimeMillis() - TIME_FRAME,
-                    TIMESTAMP_NOW);
+             contactId,
+            System.currentTimeMillis() - TIME_FRAME,
```
Comment (link2xt) [inline src/org/thoughtcrime/securesms/map/DataCollectionTask.java:86]: Something is wrong with the indentation

File: src/org/thoughtcrime/securesms/map/EmojiBitmapGenerationTask.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    super.onProgressUpdate(values);
+    EmojiBitmapGenerationCallback callback =  callbackWeakReference.get();
+    if (callback != null) {
+      callback.onEmojiBitmapCreated(values[0].first(), values[0].second());
+    }
+  }
+
+  @Override
+  protected void onPostExecute(Void aVoid) {
+    super.onPostExecute(aVoid);
```
Comment (link2xt) [inline src/org/thoughtcrime/securesms/map/EmojiBitmapGenerationTask.java:49]: Is it necessary to override these methods?

File: src/org/thoughtcrime/securesms/map/EmojiBitmapGenerationTask.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+import android.graphics.Bitmap;
+import android.os.AsyncTask;
+
+import org.thoughtcrime.securesms.components.emoji.EmojiProvider;
+import org.thoughtcrime.securesms.util.Pair;
+
+import java.lang.ref.WeakReference;
+
+public class EmojiBitmapGenerationTask extends AsyncTask<String, Pair<String, Bitmap>, Void> {
```
Comment (link2xt) [inline src/org/thoughtcrime/securesms/map/EmojiBitmapGenerationTask.java:11]: https://developer.android.com/reference/android/os/AsyncTask says AsyncTask is deprecated.

Given that progress update and other features are not used, maybe use a simple thread like suggested in https://stackoverflow.com/a/63500657 ?

File: src/org/thoughtcrime/securesms/map/MapDataManager.java [hunk 5]
```diff
@@ -279,6 +303,16 @@ public void onDataCollectionFinished() {
         }
     }
 
+    public void handleEmojiCodepoints(Set<String> emojiCodePoints) {
+      // generate only new emoji bitmaps, so check the diff
```
Comment (link2xt) [inline src/org/thoughtcrime/securesms/map/MapDataManager.java:307]: I don't understand this comment :/

**Review by link2xt — APPROVED** [review]
Seems to work, I commented on the code as much as I can.

I noticed that "Use system emoji" setting is ignored, but this is not expected, rendering system fonts into images is a lot of unnecessary additional work.

File: src/org/thoughtcrime/securesms/components/emoji/EmojiProvider.java (hunk 1 above)
Comment (cyBerta) [reply src/org/thoughtcrime/securesms/components/emoji/EmojiProvider.java:188]: the original size of each emoji is 64x64 which is far too big, if it's rendered on the map. So we need to scale the cut out emoji down.

Comment (link2xt) [reply src/org/thoughtcrime/securesms/components/emoji/EmojiProvider.java:188]: But EmojiProvider itself and getEmojiBitmap has nothing to do with the map. Maybe add an argument to set bitmap scale and pass 0.5 from the map code here?

File: src/org/thoughtcrime/securesms/map/EmojiBitmapGenerationTask.java (hunk 4 above)
Comment (cyBerta) [reply src/org/thoughtcrime/securesms/map/EmojiBitmapGenerationTask.java:11]: Dang! Thanks for the hint. It makes to not implement new stuff using the pattern.

File: src/org/thoughtcrime/securesms/map/MapDataManager.java (hunk 5 above)
Comment (cyBerta) [reply src/org/thoughtcrime/securesms/map/MapDataManager.java:307]: I'll try to explain it more in detail

File: src/org/thoughtcrime/securesms/map/EmojiBitmapGenerationTask.java (hunk 3 above)
Comment (cyBerta) [reply src/org/thoughtcrime/securesms/map/EmojiBitmapGenerationTask.java:49]: nope. thanks, that's generated boilerplate

File: src/org/thoughtcrime/securesms/components/emoji/EmojiProvider.java (hunk 1 above)
Comment (cyBerta) [reply src/org/thoughtcrime/securesms/components/emoji/EmojiProvider.java:188]: good idea +1

File: src/org/thoughtcrime/securesms/map/EmojiBitmapGenerationTask.java (hunk 4 above)
Comment (link2xt) [reply src/org/thoughtcrime/securesms/map/EmojiBitmapGenerationTask.java:11]: It's used, but only to return the result once.

Comment (cyBerta) [reply src/org/thoughtcrime/securesms/map/EmojiBitmapGenerationTask.java]: no, look at line 32, it's in a loop and it's the method to invoke onProgressUpdate()

Comment (cyBerta) [reply src/org/thoughtcrime/securesms/map/EmojiBitmapGenerationTask.java]: Anyways the whole class vanished in the meanwhile

File: src/org/thoughtcrime/securesms/components/emoji/EmojiProvider.java [hunk 6]
```diff
@@ -158,6 +172,26 @@ public int getIntrinsicHeight() {
       this.intrinsicHeight = EMOJI_RAW_HEIGHT * decodeScale;
     }
 
+    private Bitmap getEmojiBitmap(float scale) {
+      Bitmap singleEmoji = Bitmap.createBitmap((int) intrinsicWidth/2, (int) intrinsicHeight/2, Bitmap.Config.ARGB_8888);
```
Comment (link2xt) [inline src/org/thoughtcrime/securesms/components/emoji/EmojiProvider.java:176]: Shouldn't this be `intrinsicWidth * scale`? Probably makes sense to try that setting different scale (`2.5f`, `0.7f`) actually works.

Comment (cyBerta) [reply src/org/thoughtcrime/securesms/components/emoji/EmojiProvider.java:176]: +1 nice catch

Comment (cyBerta) [reply src/org/thoughtcrime/securesms/components/emoji/EmojiProvider.java]: fixed and tested with different scales

**Pull request merged**